pkg/guestagent/kubernetesservice/kubernetesservice.go

```diff
@@ -0,0 +1,130 @@
+package kubernetesservice
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type Protocol string
+
+const (
+	// UDP/SCTP when lima port forwarding works on those protocols
+	TCP Protocol = "TCP"
+)
+
+type Entry struct {
+	Protocol Protocol
+	IP       net.IP
+	Port     uint16
+}
+
+type ServiceWatcher struct {
+	serviceInformer cache.SharedIndexInformer
+}
+
+func NewServiceWatcher() *ServiceWatcher {
+	return &ServiceWatcher{}
+}
+
+func (s *ServiceWatcher) Start() {
+	retryInterval := 10 * time.Second
+	wait.PollInfinite(retryInterval, func() (done bool, err error) {
+		kubeClient, err := tryGetKubeClient()
+		if err != nil {
+			logrus.Infof("failed to get kube client: %v, will retry in %v", err, retryInterval)
+			return false, nil
+		}
+
+		ctx := context.TODO()
+		informerFactory := informers.NewSharedInformerFactory(kubeClient, time.Hour)
+		serviceInformer := informerFactory.Core().V1().Services().Informer()
+		informerFactory.Start(ctx.Done())
+		cache.WaitForCacheSync(ctx.Done(), serviceInformer.HasSynced)
+
+		s.serviceInformer = serviceInformer
+		return true, nil
+	})
+}
+
+func tryGetKubeClient() (kubernetes.Interface, error) {
+	candidateKubeConfigs := []string{
+		"/etc/rancher/k3s/k3s.yaml",
+		"/root/.kube/config",
+	}
+
+	for _, kubeconfig := range candidateKubeConfigs {
+		_, err := os.Stat(kubeconfig)
+		if err != nil {
+			if os.IsNotExist(err) {
+				continue
+			}
+
+			return nil, fmt.Errorf("stat kubeconfig %s failed: %w", kubeconfig, err)
+		}
+
+		restConfig, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return nil, fmt.Errorf("build kubeconfig from %s failed: %w", kubeconfig, err)
+		}
+		u, err := url.Parse(restConfig.Host)
+		if err != nil {
+			return nil, fmt.Errorf("parse kubeconfig host %s failed: %w", restConfig.Host, err)
+		}
+		if u.Hostname() != "127.0.0.1" { // might need to support IPv6
+			// ensures the kubeconfig points to local k8s
+			continue
+		}
+
+		kubeClient, err := kubernetes.NewForConfig(restConfig)
+		if err != nil {
+			return nil, err
+		}
+
+		return kubeClient, nil
+	}
+
+	return nil, errors.New("no valid kubeconfig found")
+}
+
+func (s *ServiceWatcher) GetPorts() []Entry {
+	if s.serviceInformer == nil {
```

Comment:
Seems to need mutex

Reply:
Yes, I use the assumption that pointer type access is atomic. But that's not great, fixed in a76ecebc212b50d43ae3441086c030920e769ed8.